source/common/http/alternate_protocols_cache_impl.cc

```diff
@@ -107,19 +107,28 @@ AlternateProtocolsCacheImpl::AlternateProtocolsCacheImpl(
     : time_source_(time_source), key_value_store_(std::move(key_value_store)),
       max_entries_(max_entries > 0 ? max_entries : 1024) {
   if (key_value_store_) {
-    KeyValueStore::ConstIterateCb load = [this](const std::string& key, const std::string& value) {
-      absl::optional<OriginData> origin_data = originDataFromString(value, time_source_, true);
-      absl::optional<Origin> origin = stringToOrigin(key);
-      if (origin_data.has_value() && origin.has_value()) {
-        setAlternativesImpl(origin.value(), origin_data.value().protocols);
-        setSrtt(origin.value(), origin_data.value().srtt);
-      } else {
-        ENVOY_LOG(warn,
-                  fmt::format("Unable to parse cache entry with key: {} value: {}", key, value));
-      }
-      return KeyValueStore::Iterate::Continue;
-    };
-    key_value_store_->iterate(load);
+    std::vector<std::pair<Origin, OriginData>> queued_origin_updates;
+
+    KeyValueStore::ConstIterateCb delayed_load =
+        [&queued_origin_updates, &time_source = time_source_](const std::string& key,
+                                                              const std::string& value) {
+          absl::optional<OriginData> origin_data = originDataFromString(value, time_source, true);
+          absl::optional<Origin> origin = stringToOrigin(key);
+          if (origin_data.has_value() && origin.has_value()) {
+            queued_origin_updates.emplace_back(origin.value(), origin_data.value());
+          } else {
+            ENVOY_LOG(warn, fmt::format("Unable to parse cache entry with key: {} value: {}", key,
+                                        value));
+          }
+          return KeyValueStore::Iterate::Continue;
+        };
+    key_value_store_->iterate(delayed_load);
+
+    // Apply updates.
+    for (auto& [origin, origin_data] : queued_origin_updates) {
+      setAlternativesImpl(origin, origin_data.protocols);
+      setSrtt(origin, origin_data.srtt);
+    }
```

Comment:
It took me a minute to figure out what's going on here. Let me make sure I've got this right. Here we are constructing a new AlternateProtocolsCache based on the serialized data in the argument key_value_store. To do this, we iterate over the key value store and parse the entries which we then use to update the information in protocols_. We update this information using setAlternativesImpl() and setSrtt() which *also* update key_value_store_. Since this happens during iteration, that's problematic.

Do I have that right?

If so, I think I'd like to propose an alternative solution. Instead of moving key_value_store into key_value_store_ in the initializer list, let's not do that. Instead, this block will iterate over the key_value_store argument (not member). Only after it is finished will it do key_value_store_ = std::move(key_value_store). 

I think this will avoid a bunch of unnecessary copies. What do you think?

Reply:
Oh sorry! The copies I was thinking about was were those involved in copying all the entries from the key value store into the queued_origin_updates and then the copies in the (redundant) addOrUpdate() on the key value store.